UndertaleModTool/MainWindow.xaml.cs

```diff
@@ -3070,127 +3115,378 @@ private async void MenuItem_OffsetMap_Click(object sender, RoutedEventArgs e)
+            {
+                Tabs[CurrentTabIndex] = newTab;
+                CurrentTabIndex = newIndex;
+            }
+
+            if (!TabController.IsLoaded)
+                CurrentTab = newTab;
+        }
+
+        public void CloseTab(bool addDefaultTab = true) // close the current tab
+        {
+            CloseTab(CurrentTabIndex, addDefaultTab);
+        }
+        public void CloseTab(int tabIndex, bool addDefaultTab = true)
+        {
+            if (tabIndex >= 0 && tabIndex < Tabs.Count)
             {
-                if (((Tab)Tabs[i]).TabTitle.Equals(newTab.TabTitle) && i != CurrentTabIndex)
+                // remove all closed tab occurrences from the selection history
+                Tab closingTab = Tabs[tabIndex];
+                for (int i = 0; i < SelectionHistory.Count; i++)
                 {
-                    CurrentTabIndex = i;
-                    tabControl.SelectedIndex = i;
+                    if (SelectionHistory[i].OpenedObject == closingTab.OpenedObject)
+                        SelectionHistory.RemoveAt(i);
+                }
+
+                // remove consecutive duplicates ( { 1, 1, 2 } -> { 1, 2 } )
+                for (int i = 0; i < SelectionHistory.Count - 1; i++)
+                {
+                    if (SelectionHistory[i] == SelectionHistory[i + 1])
+                        SelectionHistory.RemoveAt(i);
+                }
+
+                TabController.SelectionChanged -= TabController_SelectionChanged;
+
+                int currIndex = CurrentTabIndex;
+                Tabs.RemoveAt(tabIndex); // "CurrentTabIndex" changes here (bound to "TabController.SelectedIndex")
+
+                if (!closingTab.AutoClose)
+                    ClosedTabsHistory.Add(closingTab);
+
+                if (Tabs.Count == 0)
+                {
+                    CurrentTabIndex = -1;
+                    CurrentTab = null;
+
+                    if (addDefaultTab)
+                    {
+                        Highlighted = new DescriptionView("Welcome to UndertaleModTool!", "Open data.win file to get started, then double click on the items on the left to view them");
+                        OpenInTab(Highlighted);
+                        CurrentTab = Tabs[CurrentTabIndex];
+
+                        Selected = CurrentTab.OpenedObject;
+                        UpdateObjectLabel(Selected);
+                    }
+
+                    TabController.SelectionChanged += TabController_SelectionChanged;
+                }
+                else
+                {
+                    for (int i = tabIndex; i < Tabs.Count; i++)
+                        Tabs[i].TabIndex = i;
+
+                    if (currIndex == tabIndex)                            // if closing open tab
+                    {
+                        if (Tabs.Count > 1 && tabIndex < Tabs.Count - 1)  // and if that tab is not last
+                            currIndex = Tabs.Count - 1;                   // switch to the last tab
```

Comment:
Can you put the comments above the statements please, and not next to them?

Reply:
```
int currIndex = CurrentTabIndex;

// "CurrentTabIndex" changes here (bound to "TabController.SelectedIndex")
Tabs.RemoveAt(tabIndex);
```

```
// if closing open tab
if (currIndex == tabIndex)                            
{
    // and if that tab is not last
    if (Tabs.Count > 1 && tabIndex < Tabs.Count - 1)
    {
        // switch to the last tab
        currIndex = Tabs.Count - 1;
    }
    else if (currIndex != 0)
        currIndex -= 1;
```
Like that?

Reply:
From style yes.
I'd rewrite the latter comments a little:
```cs
// if closing the currently open tab
if (currIndex == tabIndex)                            
{
    // and if that tab is not the last
    if (Tabs.Count > 1 && tabIndex < Tabs.Count - 1)
    {
        // switch to the last tab
        currIndex = Tabs.Count - 1;
    }
    else if (currIndex != 0)
        currIndex -= 1;
        ```